Require instrument session to be passed when requesting a plan

callumforrester opened the pull request:
Fixes #793

callumforrester commented:
https://github.com/DiamondLightSource/blueapi/pull/1084 is a refactor that makes the auth code and this use the same cache, but can be reviewed separately if preferred

tpoliaw commented on src/blueapi/cli/cli.py:
```diff
@@ -231,6 +243,17 @@ def on_event(
+    "-i",
+    "--instrument-session",
+    type=str,
+    help=textwrap.dedent("""Instrument session associated with running the plan,
+    used to tell blueapi where to store any data and as a secuirty check:
+    the session must be valid and active and you must be a member of it.
+    If you have saved your current session with
+    blueapi controller set-instrument-session, passing this will superseed that."""),
```
```suggestion
    help=textwrap.dedent("""
        Instrument session associated with running the plan,
        used to tell blueapi where to store any data and as a security check:
        the session must be valid and active and you must be a member of it.
        If you have saved your current session with
        blueapi controller set-instrument-session, passing this will supersede that."""),
```

dedent does nothing if there's no indentation on the first line (+typos)

tpoliaw commented on src/blueapi/service/main.py:
```diff
@@ -254,23 +255,27 @@ def get_device_by_name(
     status_code=status.HTTP_201_CREATED,
 )
-@start_as_current_span(TRACER, "request", "task.name", "task.params")
+@start_as_current_span(TRACER, "request", "task_request.name", "task_request.params")
```
Would it be useful to be logging the visit as well?

tpoliaw commented on tests/conftest.py:
```diff
@@ -497,3 +498,11 @@ def mock_numtracker_server() -> Iterable[responses.RequestsMock]:
+
+
+@pytest.fixture
+def temp_dir() -> Generator[Path]:
```
There is already a `directory_path` fixture in `test_scratch` that is exactly the same as this. Could they both use the same one?

tpoliaw commented on tests/unit_tests/service/test_interface.py:
```diff
@@ -13,6 +13,7 @@
 )
 from ophyd.sim import SynAxis
 from stomp.connect import StompConnection11 as Connection
+from tests.unit_tests.service.test_rest_api import FAKE_INSTRUMENT_SESSION
```
Test modules depending on other test modules is a bit weird. Should this be a fixture or a per module constant? The `test_client` module is using literals everywhere.

tpoliaw commented on tests/unit_tests/test_cli.py:
```diff
@@ -213,26 +213,99 @@ def test_invalid_config_via_env(runner: CliRunner):
+    config_path = "tests/unit_tests/example_yaml/rest_and_stomp_config.yaml"
 
-    assert response.call_count == 1
+    mock_cache.return_value = Mock()
+    mock_cache.return_value.get.return_value = "cm12345-1"
```
```suggestion
    mock_cache().get.side_effect = {"instrument_session": "cm12345-1"}.get
```

tpoliaw commented on tests/unit_tests/test_cli.py:
```diff
@@ -213,26 +213,99 @@ def test_invalid_config_via_env(runner: CliRunner):
+    )
+
     runner.invoke(
-        main, ["-c", config_path, "controller", "run", "sleep", '{"time": 5}']
+        main,
+        ["controller", "set-instrument-session", "cm12345-1"],
```
Why is this call needed if the cache is then mocked and set directly below?

tpoliaw commented on tests/unit_tests/test_cli.py:
```diff
@@ -213,26 +213,99 @@ def test_invalid_config_via_env(runner: CliRunner):
+        ["controller", "set-instrument-session", "cm12345-1"],
+    )
+
+    assert output.stdout == "Default instrument session set to cm12345-1\n"
```
Probably also worth asserting that it was set

```
mock_cache.set.assert_called_once_with("instrument_session", "cm12345-1")
```

(or whatever the called_once_with method should be called)

tpoliaw commented on tests/unit_tests/test_cli.py:
```diff
@@ -346,14 +447,39 @@ def test_invalid_stomp_config_for_listener(runner: CliRunner):
+            '{"time": 5}',
+        ],
+    )
+    assert result.exit_code == 1
```
This test fails if the user has a visit cached. Probably want to mock some things out. Possibly with an autouse fixture to block the user's real cache being used in any of the tests.

tpoliaw commented on src/blueapi/utils/caching.py:
```diff
@@ -0,0 +1,50 @@
+        path.unlink(missing_ok=True)
+        with path.open("xb") as writer:
+            as_json = TypeAdapter(T).dump_json(value)
+            writer.write(base64.b64encode(as_json))
```
what's the benefit in base64 encoding everything?

tpoliaw commented on src/blueapi/utils/caching.py:
```diff
@@ -0,0 +1,50 @@
+    def set(self, key: str, value: Any) -> None:
+        self._ensure_root()
+        path = self._path_to_value(key)
+        path.unlink(missing_ok=True)
```
Why unlink the file first instead of opening with `wb`?

tpoliaw commented on src/blueapi/utils/caching.py:
```diff
@@ -0,0 +1,50 @@
+        os.makedirs(self._path, exist_ok=True)
+
+    def _path_to_value(self, key: str) -> Path:
+        return self._path / key
```
should probably do some kind of sanitisation so that you can't do something like `cache.set('../../file_in_users_home', 'value')`

tpoliaw reviewed (changes requested):
The changes to require a visit when submitting the plan look ok.

I'm not sure all the caching stuff is really needed and if it is, it should probably be its own PR added separately.

callumforrester commented on src/blueapi/worker/task.py:
```diff
@@ -19,18 +19,25 @@ class Task(BlueapiBaseModel):
     params: Mapping[str, Any] = Field(
         description="Values for parameters to plan, if any", default_factory=dict
     )
+    metadata: Mapping[str, Any] = Field(
+        description="Any metadata to apply to all runs within this task",
+        default_factory=dict,
+    )
```
Decided not important offline

callumforrester commented:
@tpoliaw and @DiamondJoseph I have moved the caching changes into https://github.com/DiamondLightSource/blueapi/pull/1062 so they don't block merge here. I think I've addressed all points that weren't related to that.

tpoliaw commented on src/blueapi/cli/cli.py:
```diff
@@ -231,6 +232,19 @@ def on_event(
+    "-i",
+    "--instrument-session",
+    type=str,
+    help=textwrap.dedent("""
+        Instrument session associated with running the plan,
+        used to tell blueapi where to store any data and as a security check:
+        the session must be valid and active and you must be a member of it.
+        If you have saved your current session with
+        blueapi controller set-instrument-session,
+        passing this will supersede that."""),
```
```suggestion
    help=textwrap.dedent("""
        Instrument session associated with running the plan,
        used to tell blueapi where to store any data and as a security check:
        the session must be valid and active and you must be a member of it."""),
```

tpoliaw commented on tests/unit_tests/service/test_authentication.py:
```diff
@@ -131,9 +131,9 @@ def test_processes_valid_token(
-    Path(cache._file_path).touch()
-    assert os.path.isfile(cache._file_path)
-    assert cache._file_path == f"{tmp_path}/blueapi_cache"
+def test_session_cache_manager_returns_writable_file_path(tmp_path: Path):
```
Good change but not for this PR

tpoliaw commented on tests/unit_tests/service/test_rest_api.py:
```diff
@@ -310,15 +326,23 @@ def test_get_tasks(mock_runner: Mock, client: TestClient) -> None:
+                "task": {
+                    "name": "sleep",
+                    "params": {"time": 0.0},
+                    "metadata": {},
```
Would these retain the visit when called against a real server?

tpoliaw commented on tests/unit_tests/test_cli.py:
```diff
@@ -346,14 +398,39 @@ def test_invalid_stomp_config_for_listener(runner: CliRunner):
+    assert result.exit_code == 2
+    assert "Error: Missing option '-i' / '--instrument-session'.\n" in result.stderr
+
+
```
Could use a test to cover the environment variable equivalent

```python
def test_instrument_session_from_environment(runner: CliRunner):
    with patch.dict(os.environ, {"BLUEAPI_CONTROLLER_RUN_INSTRUMENT_SESSION": "cm12345-1"}, clear=True):
        # assert visit passed to rest
```

tpoliaw reviewed (commented):
Looks ok to me. Will let Joseph look at it again before approving it
